jaxopt/_src/iterative_refinement.py

```diff
@@ -0,0 +1,210 @@
+    tol: absolute tolerance for stoping criterion (default: 1e-7).
+    verbose: If verbose=1, print error at each iteration.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled iterations.
+    implicit_diff_solve: the linear system solver to use.
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto")
+
+  References:
+    J. H. Wilkinson. Rounding Errors in Algebraic Processes. Prentice Hall, Englewood Cliffs, NJ, 1963.
+    Moler, C.B., 1967. Iterative refinement in floating point. Journal of the ACM (JACM), 14(2), pp.316-321.
+  """
+  matvec_A: Optional[Callable] = None
+  matvec_A_bar: Optional[Callable] = None
+  solve: Callable = partial(linear_solve.solve_gmres, ridge=1e-6)
+  maxiter: int = 10
+  tol: float = 1e-7
+  verbose: int = 0
+  implicit_diff_solve: Optional[Callable] = None
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def init_state(self,
+                 params: Any,
+                 params_A: Any,
+                 b: Any,
+                 params_A_bar: Any = None):
+    return IterativeRefinementState(
+      iter_num=0,
+      error=jnp.inf,
+      target_residuals=b,
+      init=params)
+
+  def update(self,
+             params: Any,
+             state: IterativeRefinementState,
+             params_A: Any,
+             b: Any,
+             params_A_bar: Optional[Any] = None):
+    if self._copy_params_A:
+      params_A_bar = params_A
+
+    A = self.matvec_A(params_A)
+    A_bar = self.matvec_A_bar(params_A_bar)
+
+    # TODO(lbethune): support preconditioners ?
+    # Could it be done by user with partial(solver, M=precond) ?
+    residual_sol = self.solve(A_bar, state.target_residuals, x0=state.init)
+
+    if params is None:
+      params = residual_sol
+    else:
+      params = tree_add(params, residual_sol)
+
+    target_residuals = tree_sub(b, A(params))
+    error = tree_l2_norm(target_residuals)
+
+    state = IterativeRefinementState(
+      iter_num=state.iter_num+1,
+      error=error,
+      target_residuals=target_residuals,
+      init=None)
+    
+    return base.OptStep(params, state)
+    
+  def optimality_fun(self,
+                     params: Any,
+                     params_A: Any,
+                     b: Any,
+                     params_A_bar: Optional[Any] = None):
+    del params_A_bar  # unused
+    A = self.matvec_A(params_A)
+    return tree_sub(b, A(params))
+
+  def l2_optimality_error(self,
+                          params: Any,
+                          params_A: Any,
+                          b: Any,
+                          params_A_bar: Optional[Any] = None):
+    del params_A_bar  # unused
+    return tree_l2_norm(self.optimality_fun(params, params_A, b))
+
+  def __post_init__(self):
+    # The inner solver might not support autodiff, so 
+    # we force implicit_diff in every case.
+    self.implicit_diff = True
```

Comment:
Is this necessary? (https://github.com/google/jaxopt/blob/main/jaxopt/_src/base.py#L120)

Reply:
It seems that the `run` method is wrapped only if `implicit_diff=True` but I might be wrong: https://github.com/google/jaxopt/blob/648f2f2553b1d91eaf54577ea0413ad64729089f/jaxopt/_src/base.py#L172

Reply:
`getattr(self, "implicit_diff", True)` returns `True` if "implicit_diff" is not present.

Reply:
Oh yeah I didn't realized that !